types/react/index.d.ts

```diff
@@ -184,10 +184,10 @@ declare namespace React {
     // Component API
     // ----------------------------------------------------------------------
 
-    type ReactInstance = Component<any, any> | Element;
+    type ReactInstance = Component<any> | Element;
 
     // Base component for plain JS classes
-    interface Component<P, S> extends ComponentLifecycle<P, S> { }
+    interface Component<P, S = {}> extends ComponentLifecycle<P, S> { }
```

Comment:
Maybe it would work if you changed `JSX.ElementClass` to `extend React.Component<any, any>`.